R/later.R

```diff
@@ -4,49 +4,203 @@
 
+.globals <- new.env(parent = emptyenv())
+
+#' Private event loops
+#'
+#' Normally, later uses a global event loop for scheduling and running
+#' functions. However, in some cases, it is useful to create a \emph{private}
+#' event loop to schedule and execute tasks without disturbing the global event
+#' loop. For example, you might have asynchronous code that queries a remote
+#' data source, but want to wait for a full back-and-forth communication to
+#' complete before continuing in your code -- from the caller's perspective, it
+#' should behave like synchronous code, and not do anything with the global
+#' event loop (which could run code unrelated to your operation). To do this,
+#' you would run your asynchronous code using a private event loop.
+#'
+#' \code{create_loop} creates and returns a handle to a private event loop,
+#' which is useful when for scheduling tasks when you do not want to interfere
+#' with the global event loop.
+#'
+#' \code{destory_loop} destroys a private event loop.
+#'
+#' \code{exists_loop} reports whether an event loop exists -- that is, that it
+#' has not been destroyed.
+#'
+#' \code{current_loop} returns the currently-active event loop. Any calls to
+#' \code{\link{later}()} or \code{\link{run_now}()} will use the current loop by
+#' default.
+#'
+#' \code{with_loop} evaluates an expression with a given event loop as the
+#' currently-active loop.
+#'
+#' \code{with_temp_loop} creates an event loop, makes it the current loop, then
+#' evaluates the given expression. Afterwards, the new event loop is destroyed.
+#'
+#' \code{global_loop} returns a handle to the global event loop.
+#'
+#'
+#' @param loop A handle to an event loop.
+#' @param expr An expression to evaluate.
+#' @param autorun Should this event loop automatically be run when its parent
+#'   loop runs? Currently, only FALSE is allowed, but in the future TRUE will
+#'   be implemented and the default. Because in the future the default will
+#'   change, for now any code that calls \code{create_loop} must explicitly
+#'   pass in \code{autorun=FALSE}.
+#' @rdname create_loop
+#'
+#' @export
+create_loop <- function(autorun = NULL) {
+  if (!identical(autorun, FALSE)) {
+    stop("autorun must be set to FALSE (until TRUE is implemented).")
+  }
+
+  id <- .globals$next_id
+  .globals$next_id <- id + 1L
+  createCallbackRegistry(id)
+
+  # Create the handle for the loop
+  loop <- new.env(parent = emptyenv())
+  class(loop) <- "event_loop"
+  loop$id <- id
+  lockBinding("id", loop)
+  if (id != 0L) {
+    # Automatically destroy the loop when the handle is GC'd (unless it's the
+    # global loop.) The global loop handle never gets GC'd under normal
+    # circumstances because .globals$global_loop refers to it. However, if the
+    # package is unloaded it can get GC'd, and we don't want the
+    # destroy_loop() finalizer to give an error message about not being able
+    # to destroy the global loop.
+    reg.finalizer(loop, destroy_loop)
```

Comment:
Did we really decide we were OK with this finalizer?

Reply:
What I recall is that we talked about it and eventually you said it looked OK. Do you have any particular concerns about problems that it could cause?

FWIW, we have a re-entrancy check for the C++ function called by the finalizer: https://github.com/r-lib/later/blob/5d535325735167889c606ef305e87ea2657e330e/src/later.cpp#L121-L129

Reply:
😬 We had a reentrancy check in pool too, but it still ended up causing problems. I guess this case is different because we're doing all of the finalization stuff in native code?

How about we merge this but I still get to say "I told you so" if something screws up? 😏